### ISSUE TITLE
Will you conduct experiments with THUMT?

### ISSUE THREAD

**Issue by NLPpupil** [marian-nmt/marian#192]
I trained transformer_base on marian and  [THUMT](https://github.com/thumt/THUMT) on  1 million English-Chinese corpora. The training speed of marian is 4142.29 words/s, while THUMT is 3x faster.

I wonder if THUMT is faster in training in most cases, will you conduct thorough experiments to compare the efficiency of marian with THUMT?

**Comment by emjotde** [comment]
Can you provide us with the command line you used for Marian? Also we have a version of transformer training that is a good deal faster. Will update in a view days.

**Comment by NLPpupil** [comment]
It is the transformer run-me.sh in examples:

`MARIAN/build/marian \
          --model model/model.npz --type transformer \
          --train-sets mydata/train.32k.en mydata/train.32k.cn \
          --max-length 100 \
          --vocabs model/vocab.ende.yml model/vocab.ende.yml \
          --mini-batch-fit -w 1000 --maxi-batch 100 \
          --early-stopping 10 --cost-type=ce-mean-words \
          --valid-freq 5000 --save-freq 5000 --disp-freq 500 \
          --valid-metrics ce-mean-words perplexity translation \
          --valid-sets mydata/valid.32k.en mydata/valid.32k.cn \
          --valid-script-path ./scripts/validate.sh \
          --valid-translation-output mydata/valid.bpe.en.output --quiet-translation \
          --valid-mini-batch 32 \
          --beam-size 5 --normalize 0.6 \
          --log model/train.log --valid-log model/valid.log \
          --enc-depth 6 --dec-depth 6 \
          --transformer-heads 8 \
          --transformer-postprocess-emb d \
          --transformer-postprocess dan \
`

**Comment by emjotde** [comment]
`-w 1000` seems wrong. You are giving it only a work space of 1GB. How much RAM does you GPU have?

**Comment by emjotde** [comment]
Also ignore the first w/s information, that one includes start-up time etc.

**Comment by NLPpupil** [comment]
Thank you!

My GPU has 10G memory, but I changed -w to 8000 got out-of-memory. Finally -w=7000 works, the training speed is 1.4x faster. Why I have 10G memory but have to set w to 7000?

Besides, the model doesn't converge, the loss keeps high at 4, what are the possible reasons?

**Comment by snukky** [comment]
Some memory is needed for the model.

The reason for bad convergence is usually suboptimal hyperparameters. If you are training on multiple GPUs, try synchronous training (`--sync-sgd`). On a single device you may consider using optimization update delay (`--optimizer-delay`) to mimic larger batches. Also, make sure you're using learning rate warm-up (`--lr-warmup`).

**Comment by emjotde** [comment]
Yes, you seem to be missing all the little quirks that are required to make the transformer model train. Do not randomly remove parameters from the example. Try this:

```
../build/marian \
    --model model/model.npz --type transformer \
    --train-sets mydata/train.32k.en mydata/train.32k.cn  \
    --max-length 100 \
    --vocabs model/vocab.ende.yml model/vocab.ende.yml \
    --mini-batch-fit -w 8000 --mini-batch 1000 --maxi-batch 1000 \
    --early-stopping 10 \
    --valid-freq 5000 --save-freq 5000 --disp-freq 500 \
    --valid-metrics cross-entropy perplexity translation \
    --valid-sets mydata/valid.32k.en mydata/valid.32k.cn \
    --valid-script-path ./scripts/validate.sh \
    --valid-translation-output mydata/valid.bpe.en.output --quiet-translation \
    --valid-mini-batch 64 \
    --beam-size 5 --normalize 0.6 \
    --log model/train.log --valid-log model/valid.log \
    --enc-depth 6 --dec-depth 6 \
    --transformer-heads 8 \
    --transformer-postprocess-emb d \
    --transformer-postprocess dan \
    --transformer-dropout 0.1 --label-smoothing 0.1 \
    --learn-rate 0.0003 --lr-warmup 16000 --lr-decay-inv-sqrt 16000 --lr-report \
    --optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \
    --tied-embeddings \
    --devices 0 --sync-sgd --seed 1111 \
    --exponential-smoothing
```

This will probably preload the whole corpus and make it faster. If you want the newer and faster transformer version try the repo https://github.com/marian-nmt/marian-dev branch `fixbeam`. I will update the main repo soon.

**Comment by NLPpupil** [comment]
@snukky Thank you. I added --optimizer-delay and --lr-warmup=4000, then the model convergences and got a good BLEU score.

**Comment by NLPpupil** [comment]
@emjotde Alright I will wait the update.
